cmd/erasure-healing_test.go

```diff
@@ -668,3 +670,88 @@ func TestHealEmptyDirectoryErasure(t *testing.T) {
 		}
 	}
 }
+
+func TestHealLastDataShard(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	nDisks := 16
+	fsDirs, err := getRandomDisks(nDisks)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer removeRoots(fsDirs)
+
+	obj, _, err := initObjectLayer(ctx, mustGetPoolEndpoints(fsDirs...))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	bucket := "bucket"
+	object := "object"
+
+	data := make([]byte, 5*humanize.MByte)
+	_, err = rand.Read(data)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var opts ObjectOptions
+
+	err = obj.MakeBucketWithLocation(ctx, bucket, BucketOptions{})
+	if err != nil {
+		t.Fatalf("Failed to make a bucket - %v", err)
+	}
+
+	_, err = obj.PutObject(ctx, bucket, object, mustGetPutObjReader(t, bytes.NewReader(data), int64(len(data)), "", ""), opts)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	actualMd5 := md5.New().Sum(data)
+
+	z := obj.(*erasureServerPools)
+	er := z.serverPools[0].getHashedSet(object)
+
+	disks := er.getDisks()
+	distribution := hashOrder(pathJoin(bucket, object), nDisks)
+	shuffledDisks := shuffleDisks(disks, distribution)
+
+	// remove last data shard
+	err = removeAll(pathJoin(shuffledDisks[11].String(), bucket, object))
+	if err != nil {
+		t.Fatalf("Failed to delete a file - %v", err)
+	}
+	_, err = obj.HealObject(ctx, bucket, object, "", madmin.HealOpts{ScanMode: madmin.HealNormalScan})
+	if err != nil {
+		t.Fatal(err)
+	}
+
```

Comment:
Can you read the object after this heal and check if the data is intact?

Reply:
I compare md5 of the healed data with origin data to check if the data is intact.

Reply:
> I compare md5 of the healed data with origin data to check if the data is intact.

My intention is when does this get corrupted? - is it in-tact here or is it corrupted later?

Reply:
in this PR, it will not get corrupted, because i fixed this bug. But in old version, it must get corrupted after second healed, the healedDataMd5 with actualMd5 is different.

Reply:
> in this PR, it will not get corrupted, because i fixed this bug. But in old version, it must get corrupted after second healed, the healedDataMd5 with actualMd5 is different.

That is why we should add a check after each heal to validate when it gets corrupted.



>

Reply:
ok, I just misunderstood what you mean, sorry.